### ISSUE TITLE
napari-app 

### ISSUE THREAD

**Issue by sofroniewn** [napari/napari#359]
## ❓ Questions and Help

Right now pip installing napari gives you access to the napari viewer that can be launched from either a jupyter notebook or from a python script. The viewer contains the core functionality required to view and manipulate images, shapes etc. but does not contain any file I/O, drag and drop, menu bars, or infrastructure for plugins. It is also not possible to launch napari from the command line directly, nor is there yet support for a fully bundled double-clickable app. Adding support for those features is in our future, but we've had various discussions about where they should live and explorations into placing them in different repos.

One approach is that the napari viewer, located on the main napari repo remains as is, and the these features -  file I/O, menu bars, drag and drop, a console, infrastructure for plugins, cli, bundled-app, etc. - all go in a new `napari-app` repo. The advantage of this approach is that it maintains a separation between complex app code, file I/O, console, and plugin stuff from the viewer, which maintains its relatively light weight dependencies, and could be used by people who don't want those features without any additional burdens. I used to favor this approach and advocated against @jni adding CLI to the napari repo - see here #208. As time has gone on though I now favor and alternative approach.

Another approach is to include file I/O, drag and drop, console, the menu bars, the ability to add plugins, and the ability to launch from the CLI all in the current repo. The actual bundled executable which you download can live in the `napari-app` repo, but there is zero functionality added in this repo, it's just the place where the do the bundling, and the link to the bundled executable is ultimately hosted elsewhere for people to download (including the readme on the main napari page). The reasons I now favor this view are as follows:

1. If everything lives in the main napari repo you get the same functionality whether you launch napari from a python script, the jupyter notebook, the CLI, or the double click app. I worry that if we didn't have the integration I described about, we'd basically end up with different versions of napari having different amounts of features that would be completely unclear to a user. For example, image a situation where a user who normally uses the napari app is now interacting with a script someone else made that launches the napari viewer and they try doing drag and drop onto the viewer but it doesn't work. I think this would be very confusing to them, and I don't see why we'd prevent our users who like to launch from scripts from having access to plugins etc as well.

2. If we have functionality split between napari-app and napari new users and developers will be confused about where to go to for help / where to post issues / where to look in the code. It will add to our maintenance burden by now being responsible for monitoring and curating two repos. It might make installation more confusing too, i.e. you pip install napari expecting the app "functionality" and CLI, but just get the viewer.

3. It will also make our development cycle more complex as changes and releases now need to be coordinated across two repos, and basic code may end up being duplicated across the repos.

4. Users would still be able to embed the napari viewer into things without the menubars, they'd just have to deal with something with a slightly larger dependency. We could also make it possible to disable features like drag-and-drop, and the console, if they did not want to expose those features. In this way, nothing is lost by integrating all the functionality into napari apart from a slightly larger install, but if we're having problems with the install of the app we probably need to fix them regardless of if it is one repo or two repos.

These thoughts are just preliminary, and to me this is an `0.2` priority for after SciPy. I'm curious how everyone else feels about this as I know we've gone back and forth on this a fw times already. The post was motivated by a recent PR on menubars #356 from @jni.

Thoughts? @jni @royerloic @kne42 @AhmetCanSolak @kevinyamauchi

**Comment by freeman-lab** [comment]
Great topic! Generally agree with this direction, and agree with all points 1-4. 

However, an additional con of putting everything inside `napari` is the potential for bloat. One thing I currently love about `napari` is that it packs a lot of functionality while remaining really lightweight from an installation and UI perspective. 

Specifically, I'd be worried about two forms of bloat
- installation becoming bulky due to extra, tough dependencies
- the UI becoming overly complex and confusing for simpler use cases

The risk of bloat varies across the various features mentioned above, and it of course depends on the implementation -- e.g. pure python fileio vs requiring bioformats and java, a console that is hidden by default and uses the built-in repl vs something more complex.

There are also ways to minimize bloat, like optional dependencies (`dask` handles this well) and flags to show and hide UI features (sublime text is particularly good here). But it requires careful design and diligence.

A final thought: this is not a one way door. We could start putting some of this functionality into `napari-app` and move it into `napari` if it feels clean and minimal. Or we could start putting it into `napari` and move it out if it becomes too bloated.

Super curious what the rest of the team thinks!

**Comment by jni** [comment]
I originally favoured having the CLI in this repo and still do. I really like to give people the ability to `pip install napari` and then immediately type `napari myimage.tif` and see their image. As we saw on Twitter, people still don't have a good solution for this and think Fiji is overkill / not nice enough. Of course, this brings us back to bloat — we don't want to hit the point where napari is thrown in the same category as Fiji for this purpose.

Nevertheless, as shown by #356 and https://github.com/napari/napari-app/pull/1 together, working on separate repos is a bit of a pain and unnecessarily slow right for our current alpha development stage. So I suggest going with @sofroniewn's suggestion, while being mindful to stay modular so that things are easy enough to rip out when the time comes. (The work I did splitting up #208 was tedious but not hard at all — the biggest difficulty was in moving over all the infrastructure/install scripts, which is a one-time cost) If people agree, I'll move https://github.com/napari/napari-app/pull/1 to #356, which should make things easier to manage.
